comp/core/workloadmeta/collectors/internal/process/process_collector_test.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/DataDog/datadog-agent/pkg/util/flavor"
+	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
+)
+
+func TestProcessCollector(t *testing.T) {
+	originalFlavor := flavor.GetFlavor()
+	defer flavor.SetFlavor(originalFlavor)
+	flavor.SetFlavor(flavor.DefaultAgent)
+
+	configOverrides := map[string]interface{}{
+		"language_detection.enabled": true,
+	}
+
+	mockStore := fxutil.Test[workloadmetamock.Mock](t, fx.Options(
+		core.MockBundle(),
+		fx.Replace(config.MockParams{Overrides: configOverrides}),
+		fx.Supply(workloadmeta.Params{
+			AgentType: workloadmeta.NodeAgent,
+		}),
+		workloadmetafxmock.MockModuleV2(),
+	))
+
+	time.Sleep(time.Second)
+
+	processDiffCh := make(chan *processwlm.ProcessCacheDiff)
+	processCollector := &collector{
+		id:            collectorID,
+		catalog:       workloadmeta.NodeAgent,
+		processDiffCh: processDiffCh,
+		store:         mockStore,
+	}
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	go processCollector.stream(ctx)
+
+	creationTime := time.Now().Unix()
+	processDiffCh <- &processwlm.ProcessCacheDiff{
+		Creation: []*processwlm.ProcessEntity{
+			{
+				Pid:          1,
+				ContainerId:  "cid",
+				NsPid:        1,
+				CreationTime: creationTime,
+				Language:     &languagemodels.Language{Name: languagemodels.Java},
+			},
+		},
+	}
+
+	expectedProc1 := &workloadmeta.Process{
+		EntityID: workloadmeta.EntityID{
+			ID:   "1",
+			Kind: workloadmeta.KindProcess,
+		},
+		NsPid:        1,
+		ContainerID:  "cid",
+		CreationTime: time.UnixMilli(creationTime),
+		Language:     &languagemodels.Language{Name: languagemodels.Java},
+	}
+
+	assert.EventuallyWithT(t, func(c *assert.CollectT) {
+		proc, err := mockStore.GetProcess(1)
+		assert.NoError(c, err)
+		assert.Equal(c, expectedProc1, proc)
+	}, time.Second, time.Millisecond*100)
+
+	processDiffCh <- &processwlm.ProcessCacheDiff{
+		Creation: []*processwlm.ProcessEntity{
+			{
+				Pid:          2,
+				ContainerId:  "cid",
+				NsPid:        2,
+				CreationTime: creationTime,
+				Language:     &languagemodels.Language{Name: languagemodels.Python},
+			},
+		},
+		Deletion: []*processwlm.ProcessEntity{
+			{
+				Pid:          1,
+				ContainerId:  "cid",
+				NsPid:        1,
+				CreationTime: creationTime,
+				Language:     &languagemodels.Language{Name: languagemodels.Java},
+			},
+		},
+	}
+
+	expectedProc2 := &workloadmeta.Process{
+		EntityID: workloadmeta.EntityID{
+			ID:   "2",
+			Kind: workloadmeta.KindProcess,
+		},
+		NsPid:        2,
+		ContainerID:  "cid",
+		CreationTime: time.UnixMilli(creationTime),
+		Language:     &languagemodels.Language{Name: languagemodels.Python},
+	}
+
+	assert.EventuallyWithT(t, func(c *assert.CollectT) {
+		proc, err := mockStore.GetProcess(2)
+		assert.NoError(c, err)
+		assert.Equal(c, expectedProc2, proc)
+
+		_, err = mockStore.GetProcess(1)
+		assert.Error(c, err)
+	}, time.Second, time.Millisecond*100)
+
+	cancel()
```

Comment:
how about moving this defer to where you create the cancel context instead?

Reply:
Done